templates/commands/render/action_include.go

```diff
@@ -36,92 +37,110 @@ func actionInclude(ctx context.Context, inc *spec.Include, sp *stepParams) error
+			return fmt.Errorf("error making relative skip path: %w", err)
+		}
+		skip[relSkipPath] = struct{}{}
 	}
 
-	incPaths, err := processPaths(inc.Paths, sp.scope)
+	skipNow := maps.Clone(skip)
+	if fromDir == sp.templateDir {
+		// If we're copying the template root directory, automatically skip
+		// 1. spec.yaml file, because it's very unlikely that the user actually
+		// wants the spec file in the template output.
+		// 2. testdata/golden directory, this is reserved for golden test usage.
+		skipNow["spec.yaml"] = struct{}{}
+		skipNow[filepath.Join("testdata", "golden")] = struct{}{}
+	}
+
+	// During validation in spec.go, we've already enforced that either:
+	// len(asPaths) is either == 0 or == len(incPaths).
+	asPaths, err := processPaths(inc.As, sp.scope)
 	if err != nil {
 		return err
 	}
 
-	asPaths, err := processPaths(inc.As, sp.scope)
+	incPaths, err := processPaths(inc.Paths, sp.scope)
 	if err != nil {
 		return err
 	}
 
 	for i, p := range incPaths {
-		// During validation in spec.go, we've already enforced that either:
-		// len(asPaths) is either == 0 or == len(incPaths).
-		as := p.Val
-		if len(asPaths) > 0 {
-			as = asPaths[i].Val
-		}
-
-		// By default, we copy from the template directory. We also support
-		// grabbing files from the destination directory, so we can modify files
-		// that already exist in the destination.
-		fromDir := sp.templateDir
-		if inc.From.Val == "destination" {
-			fromDir = sp.flags.Dest
-		}
-		absSrc := filepath.Join(fromDir, p.Val)
-		absDst := filepath.Join(sp.scratchDir, as)
-
-		skipNow := maps.Clone(skip)
-		if absSrc == sp.templateDir {
-			// If we're copying the template root directory, automatically skip
-			// 1. spec.yaml file, because it's very unlikely that the user actually
-			// wants the spec file in the template output.
-			// 2. testdata/golden directory, this is reserved for golden test usage.
-			skipNow["spec.yaml"] = struct{}{}
-			skipNow[filepath.Join("testdata", "golden")] = struct{}{}
+		matchedPaths, err := processGlobs(ctx, []model.String{p}, fromDir)
+		if err != nil {
+			return err
 		}
 
-		if _, err := sp.fs.Stat(absSrc); err != nil {
-			if common.IsStatNotExistErr(err) {
-				return p.Pos.Errorf("include path doesn't exist: %q", p.Val)
+		for _, matchedPath := range matchedPaths {
+			relMatchedPath, err := filepath.Rel(fromDir, matchedPath.Val)
+			if err != nil {
+				return fmt.Errorf("error making relative patern matched path: %w", err)
 			}
-			return fmt.Errorf("Stat(): %w", err)
-		}
 
-		params := &common.CopyParams{
-			DryRun:  false,
-			DstRoot: absDst,
-			RFS:     sp.fs,
-			SrcRoot: absSrc,
-			Visitor: func(relToAbsSrc string, de fs.DirEntry) (common.CopyHint, error) {
-				if _, ok := skipNow[relToAbsSrc]; ok {
-					return common.CopyHint{
-						Skip: true,
-					}, nil
+			as := relMatchedPath
+			if len(asPaths) > 0 { // As provided
+				if len(matchedPaths) != 1 || matchedPath.Val != filepath.Join(fromDir, p.Val) {
```

Comment:
Maybe factor out an `isGlob()` function to try to reduce the complexity of this function a bit

Reply:
I put this logic into a helper function, I think the comment should be sufficient to understand.